### PR TITLE
GEM realistic digitiser

### PR DESCRIPTION
**Pull request by tt8888tt**
Update of the GEM digitiser
More realistic digi cluster distribution matching test beam results.
https://indico.cern.ch/event/734135/contributions/3027758/attachments/1663982/2667113/rechit_cluster_size_per_ieta.pdf

@jshlee

### CODE REVIEW

File: SimMuon/GEMDigitizer/src/GEMSimpleModel.cc [hunk 1]
```diff
@@ -270,37 +271,27 @@ std::vector<std::pair<int, int> > GEMSimpleModel::simulateClustering(const GEMEt
...
-    centralStrip = topology.channel(hit_position) + 1;
-  else
-    centralStrip = topology.channel(hit_position);
-  const GlobalPoint& pointSimHit = roll->toGlobal(hit_position);
-  const GlobalPoint& pointDigiHit = roll->toGlobal(roll->centreOfStrip(centralStrip));
-  double deltaX = pointSimHit.x() - pointDigiHit.x();
+  const LocalPoint& hit_entry(simHit->entryPoint());
+  const LocalPoint& hit_exit(simHit->exitPoint());
+
+  float hit_entry_smeardX = hit_entry.x()-fabs(CLHEP::RandGaussQ::shoot(0, resolutionX_));
```
Comment (kpedro88) [inline SimMuon/GEMDigitizer/src/GEMSimpleModel.cc]: use `std::abs()` instead of `fabs()`

File: SimMuon/GEMDigitizer/src/GEMSimpleModel.cc [hunk 2]
```diff
@@ -270,37 +271,27 @@ std::vector<std::pair<int, int> > GEMSimpleModel::simulateClustering(const GEMEt
...
+  float hit_exit_smeardX = hit_exit.x()+fabs(CLHEP::RandGaussQ::shoot(0, resolutionX_));
+  if (hit_entry.x()>hit_exit.x()) {
```
Comment (kpedro88) [inline SimMuon/GEMDigitizer/src/GEMSimpleModel.cc]: You could put the above two statements in an `else` block after this to reduce unneeded random number calls

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
please test

File: SimMuon/GEMDigitizer/src/GEMSimpleModel.cc (hunk 1 above)
Comment (tt8888tt) [reply SimMuon/GEMDigitizer/src/GEMSimpleModel.cc]: I've fixed and pushed

File: SimMuon/GEMDigitizer/src/GEMSimpleModel.cc (hunk 2 above)
Comment (tt8888tt) [reply SimMuon/GEMDigitizer/src/GEMSimpleModel.cc]: you're right, I also have fixed it and pushed

**Comment by civanch** [comment]
@tt8888tt , from the talk attached to this PR it is not clear where from additional smearing come. Is it a beam quality of the test-beam or some GEM effect which is measured in this way. For me it is not clear if the default value of smearing parameter should be taken from the test-beam study or should be set to zero. I would choose zero before confirmtion from the LHC data.

**Comment by jshlee** [comment]
The current "smearing" basically adds an additional strip based on a random number, producing a max size of 2 strips. This is very unrealistic, especially for muons traversing very parallel to the gem detector surface, like cosmics. 
We can set the smearing to 0 for the meantime, but this is a more realistic simulation compared with the current version.
Our plan is to update this with LHC data.

**Comment by civanch** [comment]
What about muons, which are nearly normal to a cell? Should they be also smeared in the same way?

**Comment by jshlee** [comment]
Yes, in that case, the entry and exit points are pretty much the same and we will be smearing by the intrinsic resolution of the gem detector.
When the entry and exit points are far away, this smearing isn't really a factor.

**Comment by civanch** [comment]
You prefer left things like it is in this PR?

**Comment by jshlee** [comment]
Yes, this is far more realistic than the current digitiser. 
We will update as soon as signal gem rechit cluster size is studied with 2018 slice test data, which should only need an update to the resolution variable.

**Comment by civanch** [comment]
+1

**Comment by kpedro88** [comment]
+1

**Comment by fabiocos** [comment]
@tt8888tt @civanch @kpedro88 I see that for 2018 also RPC plots are affected, is this expected?

**Comment by jshlee** [comment]
@fabiocos - no, RPCs should not be affected.

**Comment by fabiocos** [comment]
please test

trying again for a cross check

**Comment by kpedro88** [comment]
It could be due to a change in the random number history.

**Comment by civanch** [comment]
Of course, history is changed, because extra Gauss smearing is called.

**Comment by fabiocos** [comment]
@kpedro88 @civanch shouldn't the random number sequences of the Digitizers be independent? I understand a change in the use of the GEM sequence, but isn't the RPC code using a different instance of the generator?

**Comment by fabiocos** [comment]
@tt8888tt @civanch @kpedro88 any answer? In https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_2_X_2018-06-12-1100+23530/27065/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+HARVESTFull_2018+ALCAFull_2018/ there is no change to GEM histograms and only changes to RPC histograms. Is this expected? I would like to avoid to create another muon simulation mystery that we chase for months...

File: SimMuon/GEMDigitizer/src/GEMSimpleModel.cc [hunk 3]
```diff
@@ -270,37 +271,31 @@ std::vector<std::pair<int, int> > GEMSimpleModel::simulateClustering(const GEMEt
...
-  const GlobalPoint& pointSimHit = roll->toGlobal(hit_position);
-  const GlobalPoint& pointDigiHit = roll->toGlobal(roll->centreOfStrip(centralStrip));
-  double deltaX = pointSimHit.x() - pointDigiHit.x();
+  const LocalPoint& hit_entry(simHit->entryPoint());
+  const LocalPoint& hit_exit(simHit->exitPoint());
+
+  float hit_entry_smeardX;
+  float hit_exit_smeardX;
+  if (hit_entry.x()>hit_exit.x()) {
+    hit_entry_smeardX = hit_entry.x()+std::abs(CLHEP::RandGaussQ::shoot(0, resolutionX_));
```
Comment (kpedro88) [inline SimMuon/GEMDigitizer/src/GEMSimpleModel.cc]: I think these random calls need to be: `CLHEP::RandGaussQ::shoot(engine, 0, resolutionX_)`
in order to use the specific instance of the random engine for this module.

**Comment by kpedro88** [comment]
@fabiocos there is a separate random engine requested for GEMs in https://github.com/cms-sw/cmssw/blob/e63273dd1d440fb36fe81fc9080ffb70006427d6/IOMC/RandomEngine/python/IOMC_cff.py#L169-L175, so let's see if the possible bug I noticed in https://github.com/cms-sw/cmssw/pull/23530#discussion_r195830444 solves it.

**Comment by kpedro88** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
It looks that RPC is not not changed any more, so the correct engine seems to be used now

**Comment by kpedro88** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**